sui_core/src/transaction_input_checker.rs

```diff
@@ -89,6 +174,16 @@ pub async fn check_locks(
         return Err(SuiError::LockErrors { errors });
     }
     fp_ensure!(!all_objects.is_empty(), SuiError::ObjectInputArityViolation);
+
+    // Charge gas for reading all objects from the DB.
+    // TODO: Some of the objects may be duplicate (for batch tx). We could save gas by
+    // fetching only unique objects.
+    let total_size = all_objects
+        .iter()
+        .map(|(_, obj)| obj.object_size_for_gas_metering())
```

Comment:
This call is very very expensive since it serializes everything (was just looking at the flamegraphs). We should consider caching the size as we load from DB.

Reply:
This call doesn't actually serializes everything. It should be fairly cheap most of the time. But yeah if we have it cached when loaded from DB it would be even better and more accurate.